### PR TITLE
jdk11 always create compact Strings when possible

### PR DESCRIPTION
**Pull request by pshipton**
Issue https://github.com/eclipse-openj9/openj9/issues/19543

Some testing at https://openj9-jenkins.osuosl.org/job/Pipeline-Build-Test-Personal/529/
There was one unrelated failure, an old known issue https://github.com/eclipse-openj9/openj9/issues/16608#issuecomment-2134284248

### CODE REVIEW

File: jcl/src/java.base/share/classes/com/ibm/jit/JITHelpers.java
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/com/ibm/jit/JITHelpers.java:571]: I presume the expectation is that JIT will either recognize this method and substitute a native implementation or optimize this reasonably well: if the latter, might that not be better served by having one loop variable instead of two? Perhaps like this:
```java
		int index = start << 1;
		if (!IS_BIG_ENDIAN) {
			index += 1;
		}
		for (int end = index + (length << 1); index < end; index += 2) {
			if (array[index] != 0) {
				return false;
			}
		}
		return true;
```

File: jcl/src/java.base/share/classes/java/lang/String.java
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/java/lang/String.java:859]: I don't think we want `((start != 0) || (data.length != length * 2))` included in this condition: we want to use the compressed representation whether that's true or not.

File: jcl/src/java.base/share/classes/java/lang/String.java
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/java/lang/String.java:933]: Same comment here (remove `&& ((start != 0) || (data.length != length * 2))`).

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: jcl/src/java.base/share/classes/java/lang/String.java
Comment (pshipton) [reply jcl/src/java.base/share/classes/java/lang/String.java:859]: The idea is that if the original string isn't compressed and the size doesn't change, then the new string can't be compressed either so checking it is a waste of time. Atm I believe all the callers abide by this rule.

File: jcl/src/java.base/share/classes/com/ibm/jit/JITHelpers.java
Comment (pshipton) [reply jcl/src/java.base/share/classes/com/ibm/jit/JITHelpers.java]: I was concerned about overflow but I believe this is fine. I've made the update.

**Comment by pshipton** [comment]
@hzongaro @vijaysun-omr fyi these changes to add additional checks to create compact strings in jdk11. I've added a new JITHelper.canEncodeAsLatin1() method which the JIT may want to optimize.

File: jcl/src/java.base/share/classes/java/lang/String.java
Comment (pshipton) [reply jcl/src/java.base/share/classes/java/lang/String.java:859]: There is also an optimization in the next else case where the byte[] is shared rather than creating a copy, which we don't want to loose.

Comment (keithc-ca) [reply jcl/src/java.base/share/classes/java/lang/String.java:859]: With one exception, all uses of the affected constructors (where `length != 1`) supply `start != 0` or `data.length != length * 2`, so testing that condition again here is the waste of time.

That one exception is `replaceAll()` where the `regex` is one (perhaps escaped) character and the replacement is at most one character. If the original string was compressed, the code above (lines 849-858) reuses the array as desired. If the original string was not compressed, the new code that calls `StringUTF16.newString()` handles creating a compressed string when appropriate.

File: jcl/src/java.base/share/classes/java/lang/String.java
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/java/lang/String.java:3241]: Suggest using `0` as the dummy value on line 3223 as well.

File: jcl/src/java.base/share/classes/java/lang/String.java
Comment (pshipton) [reply jcl/src/java.base/share/classes/java/lang/String.java]: Actually I'd introduced a bug in replaceAll(), it can't call String(byte[], UTF16) because this doesn't take the new length into account. Updated as suggested.

Comment (pshipton) [reply jcl/src/java.base/share/classes/java/lang/String.java]: Added back the String(byte[], UTF16) opt, but corrected the condition when it's used.

Comment (pshipton) [reply jcl/src/java.base/share/classes/java/lang/String.java]: Removed the opts in replace/replaceAll again, since the code doesn't track if the replacement character is actually used.

**Comment by vijaysun-omr** [comment]
@0xdaryl fyi since I know you were doing a survey of JCL (and String) methods to optimize as intrinsics in the near future. We may need to add the method mentioned ` JITHelper.canEncodeAsLatin1` to that list

File: jcl/src/java.base/share/classes/java/lang/String.java
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/java/lang/String.java:919]: Why did you remove this?

File: jcl/src/java.base/share/classes/java/lang/String.java
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/java/lang/String.java:3220]: Please remove extra space before the comment.

File: jcl/src/java.base/share/classes/java/lang/String.java
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/java/lang/String.java:3252]: `replacement > 255` implies that both the regex and the replacement are each 1 character long, in which case the length can't change which implies `length == newCharIndex`.
```java
					if (replacement > 255) {
```

File: jcl/src/java.base/share/classes/java/lang/String.java
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/java/lang/String.java:6714]: Please remove the trailing whitespace here.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: jcl/src/java.base/share/classes/java/lang/String.java
Comment (pshipton) [reply jcl/src/java.base/share/classes/java/lang/String.java:919]: Since the case is not used.

Comment (pshipton) [reply jcl/src/java.base/share/classes/java/lang/String.java:919]: I removed the case from the compressed branch as well, made the other review changes, and removed a few blank lines.

**Comment by pshipton** [comment]
@0xdaryl it would be good to get this in sooner rather than later, since it fixes a blocker for 0.46. I put you down as a reviewer to look at the JIT side of things.

**Comment by 0xdaryl** [comment]
The linked issues don't say what the problem is, only that it can be worked around by disabling compact Strings.  Presumably, there are situations where we assumed a compact String would be created by wasn't, and this PR fills in those gaps?  If not, please describe.

Do any performance runs need to be completed before this is merged?

Some platforms may be able to optimize the JitHelpers capability check, but depending on the perf impact it might not be possible for 0.46.  It isn't clear to me whether the JIT should does anything with the `IS_BIG_ENDIAN` check, and that might be the minimum thing we do for 0.46.  A significant performance drop attributed to `JitHelpers.canEncodeAsLatin1` would motivate doing more though.

**Comment by pshipton** [comment]
> there are situations where we assumed a compact String would be created but wasn't

Yes, correct. Also the updated behavior matches OpenJDK.

> Do any performance runs need to be completed before this is merged?

I expect there will be a perf hit when non-Latin1 characters are involved, in particular since substring is going to check all the characters. Not sure if our benchmarks test with non-Latin1 characters. Since this changes behavior to be correct, we could merge and then see if anything is exposed in the regular perf runs, rather than holding off the merge.

**Comment by keithc-ca** [comment]
> linked issues don't say what the problem is

A major issue highlighted in https://github.com/ibmruntimes/openj9-openjdk-jdk22/issues/60 is that `String.indexOf(String)` (and other API built upon it) don't work properly.

@pshipton Did you also verify that this fixes `lastIndexOf(String)`?

**Comment by pshipton** [comment]
I should also point out that jdk17 has this behavior already, since it uses the OpenJDK String impl.

**Comment by pshipton** [comment]
> Did you also verify that this fixes lastIndexOf(String)?

No, but if it doesn't work I expect it's due to different reasons than why indexOf() doesn't work. I can add similar test cases for lastIndexOf().

**Comment by pshipton** [comment]
A did a quick dup of the test cases to use lastIndexOf() instead of indexOf(). They all fail on an existing JVM, and pass with this fix. I could add them to the PR, but not sure there is much benefit as the problem is related to not creating compact strings, not the index methods.

**Comment by keithc-ca** [comment]
> the problem is related to not creating compact strings, not the index methods

That's what I expected: I agree the tests don't need to be expanded.

**Review by 0xdaryl — APPROVED** [review]

**Comment by keithc-ca** [comment]
Jenkins test sanity zlinux jdk11

**Pull request merged**
